sklearn/utils/extmath.py

```diff
@@ -822,12 +823,19 @@ def softmax(X, copy=True):
     out : ndarray of shape (M, N)
         Softmax function evaluated at every point in x.
     """
+    xp, is_array_api = get_namespace(X)
     if copy:
-        X = np.copy(X)
-    max_prob = np.max(X, axis=1).reshape((-1, 1))
+        X = xp.asarray(X, copy=True)
+    max_prob = xp.reshape(xp.max(X, axis=1), (-1, 1))
     X -= max_prob
-    np.exp(X, X)
-    sum_prob = np.sum(X, axis=1).reshape((-1, 1))
+
+    if is_array_api:
+        # array_api does not have `out=`
```

Comment:
Is there any plan or discussion for allowing this?

Reply:
No there isn't. The reason is twofold:
1. `out=` doesn't make sense for all libraries - for example, JAX and TensorFlow have immutable data structures.
2. Even for libraries that do have mutable arrays, `out=` is not a very nice API pattern. It lets users do manual optimizations that a compiler may be able to do better. There was also another argument and an alternative design presented in https://github.com/data-apis/consortium-feedback/discussions/5.

And maybe (3), IIRC NumPy and PyTorch semantics for `out=` aren't identical.

Reply:
Thanks for the feedback. I guess we will have to keep on using those `if is_array_api` conditions to protect our use of numpy's `out=` arguments for now.

I don't necessarily see that as a block for the adoption of Array API in scikit-learn, but it does make the code looks uglier... I don't really see a potential longterm fix for this.

Reply:
Here's what I get on my machine:


``` python
In [19]: def f2():
    ...:     a = np.random.rand(10000, 100000)
    ...:     a = np.exp(a)
    ...:     return a
    ...: 

In [20]: def f1():
    ...:     a = np.random.rand(10000, 100000)
    ...:     np.exp(a, out=a)
    ...:     return a
    ...: 
In [23]: %timeit f1()
15.6 s ± 2.53 s per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [24]: %timeit f2()
[1]    210906 killed     ipython
```

so the difference is quite significant (one of them gets killed 😁 )

Reply:
> Does python actually do such optimizations?

No it doesn't - but Python doesn't have a compiler? I meant a JIT or AOT compiler like JAX's JIT or PyTorch's Torchscript. It is not strange to say that in principle `X = xp.exp(X)` can be rewritten to an inplace update (i.e., `exp(X, out=X)`) by a compiler transformation if and only if the memory backing `X` isn't used elsewhere, right? 

This code looks fishy by the way for `copy=False`; the docs say nothing about inplace updating of the input array, which I'd consider a bug in library code if it were public. And to avoid this footgun, it defaults to `copy=True` which is always slow?

> (one of them gets killed grin )

Ouch, that doesn't look like the expected result.

Reply:
> No it doesn't - but Python doesn't have a compiler? I meant a JIT or AOT compiler like JAX's JIT or PyTorch's Torchscript. It is not strange to say that in principle `X = xp.exp(X)` can be rewritten to an inplace update (i.e., `exp(X, out=X)`) by a compiler transformation if and only if the memory backing `X` isn't used elsewhere, right? 

Ok now that makes sense. But here's the issue. One benefit of using array api is that developers can learn that instead of numpy's API and develop their estimators. It would also make it really easy to support array api even if they don't explicitly do so. But that raises the issue that in order to write efficient numpy code, one needs to do numpy, and for all other backends do array api, like here. I think in an ideal world, we wouldn't want to have these separate branches for the two APIs, do we?

Reply:
> I think in an ideal world, we wouldn't want to have these separate branches for the two APIs, do we?

I think we indeed want to have separate branches in as few places as possible. A lot of the places that are being identified are functions that can be added to either the array API standard (e.g., `take`, `moveaxis`) or NumPy (e.g., `unique_*`). There's a few things that will remain though, because there's an inherent tension between portability and performance. `out=` and `order=` are the two that come to mind here. And some forms of (advanced) indexing.

We reviewed the use of `out=` and `order=`, and they're used less commonly then one would expect based on the amount of discussion around them. The amount of branching that will remain in the end is quite limited, and seems like a reasonable price to pay.

Reply:
This PR now has a private helper (`_asarray_with_order`) to explicitly deal with the case where we want to enforce a specific order for numpy (e.g. when used in conjunction with Cython code) vs array API passthrough.